### PR TITLE
Jarry/ell 87 fix rust core sdk unit conversions

### PR DESCRIPTION
**Pull request by jarry-xiao**

### CODE REVIEW

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 1]
```diff
@@ -261,117 +306,81 @@ impl SDKClientCore {
...
             ..
         } = fill;
-        self.order_to_quote_amount(market_key, base_lots, price_in_ticks)
+        self.order_to_quote_atoms(market_key, base_lots, price_in_ticks)
     }
 
-    /// Takes in tick price and base lots of an order converts it into the equivalent quote amount
-    pub fn order_to_quote_amount(
+    /// Takes in tick price and base lots of an order converts it into the equivalent quote atoms
+    pub fn order_to_quote_atoms(
```
Comment (jarry-xiao) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:327]: Don't love this function name. Should represent the number of quote atoms either to transact with the order (if offer) or is locked in the book (if bid)

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 2]
```diff
@@ -166,90 +171,130 @@ impl SDKClientCore {
...
+        &self,
+        market_key: &Pubkey,
+        base_atoms: u64,
+    ) -> Result<u64> {
+        let metadata = self
             .markets
             .get(market_key)
             .ok_or_else(|| anyhow!("Market not found! Please load in the market first."))?;
+        let base_lots = base_atoms / metadata.base_atoms_per_base_lot;
+        Ok(base_lots.into())
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:197]: why does this need `.into`?

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 3]
```diff
@@ -166,90 +171,130 @@ impl SDKClientCore {
...
+        &self,
+        market_key: &Pubkey,
+        base_atoms: u64,
+    ) -> Result<u64> {
+        let metadata = self
+            .markets
+            .get(market_key)
+            .ok_or_else(|| anyhow!("Market not found! Please load in the market first."))?;
+        let base_lots = base_atoms.saturating_sub(1) / metadata.base_atoms_per_base_lot;
+        Ok(1 + base_lots)
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:211]: prefer to one-line this

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 4]
```diff
@@ -166,90 +171,130 @@ impl SDKClientCore {
...
     /// to 300_000 Widget atoms.
     pub fn base_lots_to_base_atoms(&self, market_key: &Pubkey, base_lots: u64) -> Result<u64> {
-        let market = self
+        let metadata = self
             .markets
             .get(market_key)
             .ok_or_else(|| anyhow!("Market not found! Please load in the market first."))?;
-        Ok(base_lots * market.base_lot_size) // Lot size is the number of atoms in a lot
+        let base_atoms: BaseAtoms =
+            BaseLots::new(base_lots) * BaseAtomsPerBaseLot::new(metadata.base_atoms_per_base_lot);
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:224]: prefer to not use formal units here for consistency

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 5]
```diff
@@ -166,90 +171,130 @@ impl SDKClientCore {
...
     /// convert 3 USDC to quote lots you would call sdk.quote_unit_to_quote_lots(3.0). This would return
     /// the number of quote lots that would be equivalent to 3 USDC.
     pub fn quote_units_to_quote_lots(&self, market_key: &Pubkey, quote_units: f64) -> Result<u64> {
-        let market = self
+        let metadata = self
             .markets
             .get(market_key)
             .ok_or_else(|| anyhow!("Market not found! Please load in the market first."))?;
-        Ok((quote_units * market.quote_multiplier as f64 / market.quote_lot_size as f64) as u64)
+        Ok((quote_units * metadata.quote_atoms_per_quote_unit as f64
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:237]: as above, prefer to do `quote_atoms_per_quote_unit / quote_atoms_per_quote_lot * quote_units` for smaller intermediate values. first division should always be integer so no precision lost

we should have rounding for this one too (to be consistent w/ the raw base units to base lots)

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 6]
```diff
@@ -166,90 +171,130 @@ impl SDKClientCore {
...
+        market_key: &Pubkey,
+        quote_atoms: u64,
+    ) -> Result<u64> {
+        let metadata = self
             .markets
             .get(market_key)
             .ok_or_else(|| anyhow!("Market not found! Please load in the market first."))?;
-        Ok(quote_atoms / market.quote_lot_size)
+        let quote_lots = quote_atoms / metadata.quote_atoms_per_quote_lot;
+        Ok(quote_lots.into())
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:255]: is `.into()` necessary?

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 7]
```diff
@@ -166,90 +171,130 @@ impl SDKClientCore {
...
+    }
+
+    pub fn quote_atoms_to_quote_lots_rounded_up(
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:259]: doc

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 8]
```diff
@@ -166,90 +171,130 @@ impl SDKClientCore {
...
+        &self,
+        market_key: &Pubkey,
+        quote_atoms: u64,
+    ) -> Result<u64> {
+        let metadata = self
+            .markets
+            .get(market_key)
+            .ok_or_else(|| anyhow!("Market not found! Please load in the market first."))?;
+        let quote_lots = quote_atoms.saturating_sub(1) / metadata.quote_atoms_per_quote_lot;
+        Ok(1 + quote_lots)
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:269]: oneline

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 9]
```diff
@@ -166,90 +171,130 @@ impl SDKClientCore {
...
     /// 100 quote atoms per quote lot of USDC, you would call sdk.quote_lots_to_quote_atoms(300) to convert 300 quote lots
     /// to 30_000 USDC atoms.
     pub fn quote_lots_to_quote_atoms(&self, market_key: &Pubkey, quote_lots: u64) -> Result<u64> {
-        let market = self
+        let metadata = self
             .markets
             .get(market_key)
             .ok_or_else(|| anyhow!("Market not found! Please load in the market first."))?;
-        Ok(quote_lots * market.quote_lot_size)
+        let quote_atoms: QuoteAtoms = QuoteLots::new(quote_lots)
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:281]: why units here and nowhere else?

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 10]
```diff
@@ -166,90 +171,130 @@ impl SDKClientCore {
...
+            * QuoteAtomsPerQuoteLot::new(metadata.quote_atoms_per_quote_lot);
+        Ok(quote_atoms.into())
     }
 
     /// Converts a number of base atoms to a floating point number of base units. For example if the base currency
     /// is a Widget where the token has 9 decimals and you wanted to convert 1_000_000_000 base atoms to
     /// a floating point number of whole Widget tokens you would call sdk.base_amount_to_float(1_000_000_000). This
     /// would return 1.0. This is useful for displaying the base amount in a human readable format.
-    pub fn base_atoms_to_base_unit_as_float(
+    pub fn base_atoms_to_raw_base_unit_as_float(
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:290]: `to_raw_base_unitS`

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 11]
```diff
@@ -261,117 +306,81 @@ impl SDKClientCore {
         market_key: &Pubkey,
         quote_atoms: u64,
     ) -> Result<f64> {
-        let market = self
+        let metadata = self
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:311]: method name should be `quote_unitS`

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 1 above)
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:329]: fix doc

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 12]
```diff
@@ -261,117 +306,81 @@ impl SDKClientCore {
...
-                / market.num_base_lots_per_base_unit,
+            base_lots * price_in_ticks * metadata.tick_size_in_quote_atoms_per_base_unit
+                / metadata.num_base_lots_per_base_unit,
         )
     }
 
     /// Takes in a price as a floating point number and converts it to a number of ticks (rounded down)
-    pub fn float_price_to_ticks(&self, market_key: &Pubkey, price: f64) -> Result<u64> {
-        let market = self
+    pub fn float_price_to_ticks_rounded_down(
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:346]: note units of price in doc (quote units per raw base unit)

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 13]
```diff
@@ -261,117 +306,81 @@ impl SDKClientCore {
...
-                .ceil() as u64,
-        )
+        Ok(((price
+            * metadata.raw_base_units_per_base_unit as f64
+            * metadata.quote_atoms_per_quote_unit as f64)
+            / metadata.tick_size_in_quote_atoms_per_base_unit as f64)
+            .ceil() as u64)
     }
 
     /// Takes in a number of ticks and converts it to a floating point number price
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:374]: add units to doc

**Review by eugene-chen — COMMENTED** [review]
I think we need to bump phoenix-sdk version too

**Comment by eugene-chen** [comment]
Let's update the ESL docs that are currently on all the conversion methods

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 14]
```diff
@@ -107,29 +111,30 @@ pub struct MarketMetadata {
...
+        let base_atoms_per_base_unit = 10u64.pow(header.base_params.decimals);
         let base_mint = header.base_params.mint_key;
         let quote_mint = header.quote_params.mint_key;
         let tick_size_in_quote_atoms_per_base_unit =
             header.get_tick_size_in_quote_atoms_per_base_unit().into();
         // max(1) is only relevant for old markets where the raw_base_units_per_base_unit was not set
         let raw_base_units_per_base_unit = header.raw_base_units_per_base_unit.max(1);
-        let num_base_lots_per_base_unit =
-            base_multiplier * raw_base_units_per_base_unit as u64 / base_lot_size;
+        let num_base_lots_per_base_unit = base_atoms_per_base_unit
```
[the remark this answers is not in the document]
Comment (jarry-xiao) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:124]: i can upcast to make sure this never overflows

[the remark this answers is not in the document]
Comment (eugene-chen) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:124]: this really should never overflow

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 15]
```diff
@@ -146,232 +156,233 @@ impl SDKClientCore {
...
     /// the number of quote lots that would be equivalent to 3 USDC.
     pub fn quote_units_to_quote_lots(&self, market_key: &Pubkey, quote_units: f64) -> Result<u64> {
-        let market = self
+        let metadata = self
             .markets
             .get(market_key)
             .ok_or_else(|| anyhow!("Market not found! Please load in the market first."))?;
-        Ok((quote_units * market.quote_multiplier as f64 / market.quote_lot_size as f64) as u64)
+        Ok((quote_units
+            * (metadata.quote_atoms_per_quote_unit / metadata.quote_atoms_per_quote_lot) as f64)
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:226]: I think this should do everything as u64. `quote_atoms_per_quote_unit / quote_atoms_per_quote_lot` should always be an integer

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 1 above)
Comment (eugene-chen) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:303]: We should also refactor to either:
* actually take in some sort of order object and update doc accordingly OR
* update function name to show that this takes a base amount and a price

File: rust/phoenix-sdk-core/src/test_unit_conversion.rs [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+    let core = setup_with_raw_base_unit_multiplier(&market, 1000);
+    let base_atoms = 1000000;
+    let raw_base_unit = core
+        .base_atoms_to_raw_base_units_as_float(&market, base_atoms)
+        .unwrap();
+    assert_eq!(raw_base_unit, 0.001);
+}
+
+#[test]
+fn test_quote_atoms_to_quote_unit_as_float() {
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/test_unit_conversion.rs:220]: `to_quote_unitS`

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 15 above)
Comment (jarry-xiao) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:226]: quote_units is a float though

**Review by eugene-chen — APPROVED** [review]

**Pull request merged**